Infinispan 15: indexing and query news (#353)
infinispan/infinispan.github.io: _posts/2024-06-10-infinispan-15-queries.adoc

---
layout: blog
title: Infinispan 15 indexing & query news
permalink: /blog/:year/:month/:day/infinispan-15-queries
date: '2024-06-10T00:00:00.000-00:00'
author: fax4ever
tags: [ "search", "vector", "score", "knn", "indexing", "embeddings" ]
---

= Infinispan 15 indexing & query news

> The Infinispan team is happy to announce some improvements on query APIs that are already present on Infinispan 15:

Reviewer: @rigazilla is right. This should be more like "A short while back we released Infinispan 15 which delivered many improvements to the query API. This blog is an in-depth dive into some of these"

> The Infinispan team is happy to announce some improvements on query APIs that are already present on Infinispan 15:

Author: sure! I'm applying the changes

1. <<rest_projections,Rest queries with projections>>: more projection types are supported using the REST API.
2. <<continuous-queries,Query cache API>>: regular and continuous queries can be defined directly from the cache API.
3. <<filter-elements,Filter elements for kNN queries>>: it is now possible to filter out the set of entities on which to apply a kNN-vector search.
4. <<index-query-by-keys, Index by keys - query by keys>>: indexes can be defined also on the keys, not only on the values, so that we can query also the keys values.

[[rest_projections]]
== Rest queries with projections

Recently we introduced score and versions projections, that along with entity and field projections are the all
kinds of projections we support with the Infinispan queries.

The same projections are now available using the REST query API.

> For instance if we have an entity like the following:

Reviewer 2: I would group entity, query and result... something like `Following an example of entity, query and result:`

   @ProtoFactory
   public Game(String name, String description) {
      this.name = name;
      this.description = description;
   }

   @Keyword(projectable = true, sortable = true)
   @ProtoField(1)
   public String getName() {
      return name;
   }

   @Text
   @ProtoField(2)
   public String getDescription() {
      return description;
   }
}
----

Running the following query:

[source,sql]
----
select g, g.description, version(g), score(g) from Game g where g.description : 'bla3'
----

This is a possible outcome:

[source,json]
----
{
    "hit_count": 1,
    "hit_count_exact": true,
    "hits": [
        {
            "hit": {
                "version()": 7,
                "description": "bla bla3",
                "*": {
                    "_type": "Game",
                    "name": "bla3",
                    "description": "bla bla3"
                },
                "score()": 0.90565
            }
        }
    ]
}
----

In this case we have requested the version projection using the projection function `version(g)`
and receiving the corresponding result in the `version()` attribute of the returned `hit`.
Similarly, a `score()` projection is produced.

> The special value `*` corresponds to the entity projection result, in which the special field `_type` corresponds to the type of the entity.

Reviewer 2: ```suggestion
The special attribute `*` corresponds to the entity projection result, in which the special field `_type` corresponds to the type of the entity.
```

[[continuous-queries]]
== Query cache API

> The use of the search factory to access the query APIs is no longer required and it is now deprecated.
> 
> From Infinispan 15 we can simply create a query instance from both a remote or an embedded cache (the API is the same)
> using the method `cache#query`.

Reviewer 2: ```suggestion
With Infinispan 15, both embedded and remote caches can be queried with the same API metho `cache#query`. The use of the search factory to access the query APIs is no longer required and it is now deprecated.
```

Here is an example:

[source,java]
----
Query<Person> query = myCache.query("FROM space.Person WHERE name = 'user1' AND age > 20");
----

From this point forward we will use the usual API to configure and run the query. For instance:

[source,java]
----
query.startOffset(10);
query.maxResults(10);
QueryResult<Person> = query.execute();
----

> Similarly, it is possible to get a continuous query instance from the same cache (both remote and embedded),
> using the method `cache#continuousQuery`:

Reviewer 2: ```suggestion
Similarly, it is possible to get a continuous query instance (both for remote and embedded cache), using the method `cache#continuousQuery`:
```

[source,java]
----
ContinuousQuery<Integer, Person> continuousQuery = myCache.continuousQuery();
----

> From which as usual it will be possible to define the continuous callback:

Reviewer 2: I would remove this and group the example

Author: I would prefer to keep it, to see where the story ends, even if we didn't changed this API.

[source,java]
----
continuousQuery.addContinuousQueryListener(query, new ContinuousQueryListener<>() {

      @Override
      public void resultJoining(Integer key, Object value) {
         // handle entity creations
      }
      @Override
      public void resultUpdated(Integer key, Object value) {
         // handle entry updates
      }
      @Override
      public void resultLeaving(Integer key) {
         // handle entry leavings
      }
   });
----

[[filter-elements]]
== Filter elements for kNN queries

> kNN queries can be run filtering the population on which to apply the search.

Reviewer 2: ```suggestion
kNN queries can run on a pre-filtered subset of the population.
```

Author: I prefer the original from, I think it is a little more expressive.

> A kNN filter is defined using any kind of predicate (included boolean expressions) provided by the Infinispan queries.

Reviewer 2: ```suggestion
A kNN filter is defined using any kind of predicate (included boolean expressions) provided by the Infinispan query language.
```

For instance, let's consider the following entity:

[source,java]
----
@Indexed
public class Item {

   private String code;
   private float[] floatVector;
   private String description;

   @ProtoFactory
   public Item(String code, float[] floatVector, String description) {
      this.code = code;
      this.floatVector = floatVector;
      this.description = description;
   }

   @Keyword
   @ProtoField(1)
   public String getCode() {
      return code;
   }

   @Vector(dimension = 3)
   @ProtoField(2)
   public float[] getFloatVector() {
      return floatVector;
   }

   @Text
   @ProtoField(3)
   public String getDescription() {
      return description;
   }
}
----

> which is mapped using different indexing annotations (`@Keyword`, `@Text`, `@Vector`, ...).
> 
> We can limit the population on which to apply the search to the only items having a description text containing the term `cat`.

Reviewer 2: ```suggestion
Suppose that we want to limit the vector search only to record with the word 'cat' in the description. We can do like this:
```

[source,java]
----
Query<Object[]> query = remoteCache.query(
      "select score(i), i from Item i where i.floatVector <-> [:a]~:k filtering i.description : 'cat'");
query.setParameter("a", new float[]{7.0f, 7.0f, 7.0f});
query.setParameter("k", 3);

List<Object[]> hits = query.list();
----

> In particular, this is also an example of how to combine full text search and vector search.
> 
> Also boolean composite predicates can be used to filter the search.

Reviewer 2: ```suggestion
This is example shows a combination of full text search and vector search, but boolean composite predicates are also supported.
```

> In the following we will limit the search to the only items having the term `cat` in their description and having code `w739`.

Reviewer 2: ```suggestion
In the following we will limit the search only to the items having the term `cat` in their description and having code `w739`.
```
